File: flowlet-frontend/src/services/projectService.ts
import api from "./api";
import { Project } from "@/store/projectStore";
import type { ApiResponse } from "@/types";

/**
 * 项目成员角色
 */
export type ProjectRole = "owner" | "admin" | "editor" | "viewer";

/**
 * 项目成员信息
 */
export interface ProjectMember {
  id: string;
  projectId: string;
  userId: string;
  role: ProjectRole;
  createdAt?: string;
  updatedAt?: string;
}

/**
 * 创建项目请求
 */
export interface CreateProjectRequest {
  name: string;
  description?: string;
}

/**
 * 更新项目请求
 */
export interface UpdateProjectRequest {
  name: string;
  description?: string;
}

/**
 * 添加成员请求
 */
export interface AddMemberRequest {
  userId: string;
  role: ProjectRole;
}

/**
 * 用户权限信息
 */
export interface ProjectPermissions {
  role: ProjectRole | null;
  canEdit: boolean;
  canManage: boolean;
  isOwner: boolean;
}

/**
 * 项目管理服务
 */
export const projectService = {
  /**
   * 获取当前用户可访问的项目列表
   */
  async listProjects(): Promise<Project[]> {
    const response = await api.get<ApiResponse<Project[]>>("/projects");
    return response.data.data;
  },

  /**
   * 获取项目详情
   */
  async getProject(projectId: string): Promise<Project> {
    const response = await api.get<ApiResponse<Project>>(
      `/projects/${projectId}`
    );
    return response.data.data;
  },

  /**
   * 创建项目
   */
  async createProject(request: CreateProjectRequest): Promise<Project> {
    const response = await api.post<ApiResponse<Project>>("/projects", request);
    return response.data.data;
  },

  /**
   * 更新项目
   */
  async updateProject(
    projectId: string,
    request: UpdateProjectRequest
  ): Promise<Project> {
    const response = await api.put<ApiResponse<Project>>(
      `/projects/${projectId}`,
      request
    );
    return response.data.data;
  },

  /**
   * 删除项目
   */
  async deleteProject(projectId: string): Promise<void> {
    await api.delete<ApiResponse<void>>(`/projects/${projectId}`);
  },

  /**
   * 获取项目成员列表
   */
  async listMembers(projectId: string): Promise<ProjectMember[]> {
    const response = await api.get<ApiResponse<ProjectMember[]>>(
      `/projects/${projectId}/members`
    );
    return response.data.data;
  },

  /**
   * 添加项目成员
   */
  async addMember(
    projectId: string,
    request: AddMemberRequest
  ): Promise<ProjectMember> {
    const response = await api.post<ApiResponse<ProjectMember>>(
      `/projects/${projectId}/members`,
      request
    );
    return response.data.data;
  },

  /**
   * 更新成员角色
   */
  async updateMemberRole(
    projectId: string,
    userId: string,
    role: ProjectRole
  ): Promise<ProjectMember> {
    const response = await api.put<ApiResponse<ProjectMember>>(
      `/projects/${projectId}/members/${userId}`,
      { role }
    );
    return response.data.data;
  },

  /**
   * 移除项目成员
   */
  async removeMember(projectId: string, userId: string): Promise<void> {
    await api.delete<ApiResponse<void>>(
      `/projects/${projectId}/members/${userId}`
    );
  },

  /**
   * 获取当前用户在项目中的权限
   */
  async getMyPermissions(projectId: string): Promise<ProjectPermissions> {
    const response = await api.get<ApiResponse<ProjectPermissions>>(
      `/projects/${projectId}/permissions`
    );
    return response.data.data;
  },
};

export default projectService;

File: flowlet-backend/src/main/java/com/flowlet/service/ProjectService.java
package com.flowlet.service;

import com.flowlet.entity.Project;
import com.flowlet.entity.ProjectMember;

import java.util.List;

/**
 * 项目服务接口
 */
public interface ProjectService {

    /**
     * 创建项目
     * 创建者自动成为 owner
     */
    Project create(String name, String description);

    /**
     * 获取项目详情
     */
    Project getById(String projectId);

    /**
     * 获取当前用户可见的项目列表
     */
    List<Project> listMyProjects();

    /**
     * 更新项目信息
     */
    Project update(String projectId, String name, String description);

    /**
     * 删除项目
     * 只有 owner 可以删除
     */
    void delete(String projectId);

    /**
     * 添加项目成员
     */
    ProjectMember addMember(String projectId, String userId, String role);

    /**
     * 移除项目成员
     */
    void removeMember(String projectId, String userId);

    /**
     * 更新成员角色
     */
    ProjectMember updateMemberRole(String projectId, String userId, String role);

    /**
     * 获取项目成员列表
     */
    List<ProjectMember> listMembers(String projectId);

    /**
     * 检查用户是否有项目访问权限
     */
    boolean hasAccess(String projectId, String userId);

    /**
     * 检查用户是否有项目编辑权限
     */
    boolean canEdit(String projectId, String userId);

    /**
     * 检查用户是否有项目管理权限
     */
    boolean canManage(String projectId, String userId);

    /**
     * 获取用户在项目中的角色
     */
    ProjectMember.Role getUserRole(String projectId, String userId);
}
